src/Spectre.libDataset/Dataset.h

```diff
@@ -0,0 +1,192 @@
+
+        /// <summary>
+        /// Gets the dataset metadata in read-only fashion.
+        /// </summary>
+        /// <returns>Dataset metadata</returns>
+        const DatasetMetadata& getDatasetMetadata() const override
+        {
+            return m_metadata;
+        }
+
+        /// <summary>
+        /// Gets sample under specified index.
+        /// </summary>
+        /// <param name="idx">The index.</param>
+        /// <returns>Sample</returns>
+        DataType& operator[](size_t idx) override
+        {
+            if (idx < m_data.size())
+            {
+                return m_data[idx];
+            }
+            else
+            {
+                throw OutOfRange(idx, m_data.size());
+            }
+        }
+
+        /// <summary>
+        /// Gets the sample metadata.
+        /// </summary>
+        /// <param name="idx">The index.</param>
+        /// <returns>Sample metadata</returns>
+        SampleMetadata& getSampleMetadata(size_t idx) override
+        {
+            if (idx < m_sampleMetadata.size())
+            {
+                return m_sampleMetadata[idx];
+            }
+            else
+            {
+                throw OutOfRange(idx, m_sampleMetadata.size());
+            }
+        }
+
+        /// <summary>
+        /// Gets the dataset metadata.
+        /// </summary>
+        /// <returns>Dataset metadata</returns>
+        DatasetMetadata& getDatasetMetadata() override
+        {
+            return m_metadata;
+        }
+
+        /// <summary>
+        /// Gets the data in read-only fashion.
+        /// </summary>
+        /// <returns></returns>
+        gsl::span<const DataType> getData() const override
+        {
+            return gsl::span<const DataType>(m_data);
+        }
+
+        /// <summary>
+        /// Gets the sample metadata in read-only fashion.
+        /// </summary>
+        /// <returns></returns>
+        gsl::span<const SampleMetadata> getSampleMetadata() const override
+        {
+            return gsl::span<const SampleMetadata>(m_sampleMetadata);
+        }
+
+        /// <summary>
+        /// Adds the sample.
+        /// </summary>
+        /// <param name="sample">The sample.</param>
+        /// <param name="metadata">The metadata.</param>
+        void addSample(DataType sample, SampleMetadata metadata) override
+        {
+            m_data.push_back(sample);
+            m_sampleMetadata.push_back(metadata);
+        }
+
+        /// <summary>
+        /// Number of elements in dataset.
+        /// </summary>
+        /// <returns>Size</returns>
+        size_t size() const override
+        {
+            return m_data.size();
+        }
+
+        /// <summary>
+        /// Checks, whether dataset is empty.
+        /// </summary>
+        /// <returns>True, if dataset is empty.</returns>
+        bool empty() const override
+        {
+            return m_data.empty();
+        }
+
+    private:
+        std::vector<DataType> m_data;
```

Comment:
According to [C++ coding convetion](https://github.com/spectre-team/spectre/wiki/Cpp-coding-convention) class members should start with m_ (it's ok) and follow up with an uppercase letter.